### PR TITLE
EDU-1034 Simplify registration workflow

### PR DESCRIPTION
**Pull request by ahelmberger**
Closes https://educandu.atlassian.net/browse/EDU-1034

### CODE REVIEW

File: src/components/registration-form.js [hunk 1]
```diff
@@ -1,148 +1,285 @@
...
+  };
 
-  const register = async ({ email, password, displayName }) => {
+  const handleEnterDataFinish = async ({ email, password, displayName }) => {
     try {
-      onRegistrationStarted();
-      const { result, user: registeredUser } = await userApiClient.register({ email, password, displayName });
+      setIsLoading(true);
+      const { result, user } = await userApiClient.requestRegistration({
+        email,
```
Comment (adrianaluchian) [inline src/components/registration-form.js:106]: Should we trim this guy too? 🤔 We do it at login but not here. Or we could maybe trim both on the server side actually, might be better as we lowercase the email there anyway.

File: src/components/registration-form.js [hunk 2]
```diff
@@ -1,148 +1,285 @@
...
-    formRef.current = form;
-  }
+  useEffect(() => {
+    if (currentStepKey === STEP.enterData) {
+      setCreatedUser(null);
+      setForbiddenEmails([]);
+      setInvalidVerificationCodes([]);
+    }
+    setSteps(previousSteps => {
+      const newSteps = currentStepKey === STEP.enterData ? createDefaultSteps() : previousSteps.map(step => ({ ...step }));
```
Comment (adrianaluchian) [inline src/components/registration-form.js:84]: For the sake of better clarity I would suggest to create a `setStepProcess` method just like `setStepError` and `setStepSuccess`. That pattern makes it so much easier to follow these state changes.

File: src/components/registration-form.js [hunk 3]
```diff
@@ -1,148 +1,285 @@
...
+  const renderEnterDataForm = () => {
+    const termsPage = settings.termsPage?.[uiLanguage];
+    const agreementValidationRules = [
+      {
+        message: t('confirmTerms'),
+        validator: (_, value) => value ? Promise.resolve() : Promise.reject(new Error(t('confirmTerms')))
+      }
+    ];
+    return (
+      <div className="RegistrationForm-form">
```
Comment (adrianaluchian) [inline src/components/registration-form.js:169]: Small thing and the code is old, but we could drop the `-form` in the class name, as it's the outer-most class in this component.

File: src/components/registration-form.js [hunk 4]
```diff
@@ -1,148 +1,285 @@
...
+      {
+        validator: (_rule, value) => {
+          return value && invalidVerificationCodes.includes(value)
+            ? Promise.reject(new Error(t('verificationCodeIsInvalid')))
+            : Promise.resolve();
+        }
+      }
+    ];
+    return (
+      <div className="RegistrationForm-form">
```
Comment (adrianaluchian) [inline src/components/registration-form.js:221]: Here as well.

File: src/components/registration-form.js [hunk 5]
```diff
@@ -1,148 +1,285 @@
...
+  };
+
   return (
-    <div className="RegistrationForm">
-      {user ? registrationConfirmation : registrationForm}
+    <div className="RegistrationForm u-panel">
+      {renderSteps()}
+      {currentStepKey === STEP.enterData && renderEnterDataForm()}
+      {currentStepKey === STEP.enterCode && renderEnterCodeForm()}
+      {currentStepKey === STEP.completed && renderCompletedMessage()}
```
Comment (adrianaluchian) [inline src/components/registration-form.js:280]: Nice switching mechanism with these steps and states! 👍

File: src/components/pages/complete-password-reset.js [hunk 6]
```diff
@@ -63,7 +62,6 @@ function CompletePasswordReset({ initialState, PageTemplate, SiteLogo }) {
     <div className="CompletePasswordResetPage-form">
       <Form onFinish={handleFinish} scrollToFirstError>
         <PasswordFormItem name="password" {...formItemLayout} />
-        <PasswordConfirmationFormItem name="confirm" passwordFormItemName="password" {...formItemLayout} />
```
Comment (adrianaluchian) [inline src/components/pages/complete-password-reset.js:66]: 👏 💯

File: src/server/user-controller.spec.js [hunk 7]
```diff
@@ -219,6 +222,119 @@ describe('user-controller', () => {
...
+            headers: { host: 'localhost' },
+            session: { externalAccount: null },
+            body: { userId: 'cnztc5ztm41mx03z', verificationCode: '3xzrxzt43z0xtm1' },
+            login: sandbox.stub().callsArg(1)
+          });
+          res = httpMocks.createResponse({ eventEmitter: events.EventEmitter });
+
+          res.on('end', resolve);
+
+          userService.verifyUser.resolves(dbUser);
```
Comment (adrianaluchian) [inline src/server/user-controller.spec.js:245]: Just a tiny thing, but for extra safety, we could set this up as `userService.verifyUser.withArgs(req.body.userId, req.body.verificationCode).resolves(dbUser)`

**Comment by adrianaluchian** [comment]
This looks absolutely wonderful now! 😍 Makes up for such a nicer experience and is much more clear for the user! Was a good idea to change and it's nicely implemented.

File: src/components/registration-form.js (hunk 1 above)
Comment (ahelmberger) [reply src/components/registration-form.js:106]: I trimmed it here as I consider it more a UI concern than the lowercasing, so the API gets already sanitized values.

File: src/components/registration-form.js (hunk 2 above)
Comment (ahelmberger) [reply src/components/registration-form.js:84]: Good idea, simplified the function accordingly.

File: src/components/pages/complete-password-reset.js (hunk 6 above)
Comment (ahelmberger) [reply src/components/pages/complete-password-reset.js:66]: Yeah, I liked that too :-)

File: src/components/registration-form.js (hunk 5 above)
Comment (ahelmberger) [reply src/components/registration-form.js:275]: :+1:

**Review by adrianaluchian — APPROVED** [review]

**Pull request merged**